File: matlab_sim/sinr_rate.m
function [gamma, rates, sum_rate] = sinr_rate(cfg, h_eff, p_dbw)
%SINR_RATE Compute SINR and rates.

num_users = size(h_eff, 2);

p_watts = 10.^(p_dbw / 10);
p_k = p_watts / num_users;

w = zeros(cfg.nt, num_users);
for k = 1:num_users
    hk = h_eff(:, k);
    w(:, k) = hk / (norm(hk) + cfg.eps);
end

signal = zeros(num_users, 1);
interference = zeros(num_users, 1);

for k = 1:num_users
    hk = h_eff(:, k);
    for i = 1:num_users
        gain = abs(hk' * w(:, i)).^2;
        if i == k
            signal(k) = signal(k) + p_k * gain;
        else
            interference(k) = interference(k) + p_k * gain;
        end
    end
end

gamma = signal ./ (interference + cfg.noise_watts + cfg.eps);

rates = cfg.bandwidth * log2(1 + gamma);
sum_rate = sum(rates);
end

File: sim/sinr_rate.py
"""Compute SINR and rates for multi-user transmission."""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np


@dataclass
class RateResult:
    sinr: np.ndarray
    rates: np.ndarray
    sum_rate: float
    beamformers: np.ndarray
    power_alloc: np.ndarray


def compute_rates(
    h_eff: np.ndarray,
    total_power: float,
    noise_power: float,
    bandwidth: float,
) -> RateResult:
    """Compute multi-user SINR and rates with MRT beamforming."""
    k_total, nt = h_eff.shape
    power_alloc = np.full(k_total, total_power / k_total)

    norms = np.linalg.norm(h_eff, axis=1)
    norms = np.where(norms > 0, norms, 1.0)
    beamformers = h_eff / norms[:, None]

    gains = np.abs(h_eff @ beamformers.conj().T) ** 2
    signal = power_alloc * np.diag(gains)
    interference = gains @ power_alloc - signal
    sinr = signal / (interference + noise_power)
    rates = bandwidth * np.log2(1.0 + sinr)

    return RateResult(
        sinr=sinr,
        rates=rates,
        sum_rate=float(np.sum(rates)),
        beamformers=beamformers,
        power_alloc=power_alloc,
    )
